### ISSUE TITLE
calculate_outcomes, bin='all': all the input arrays must have same number of dimensions

### ISSUE THREAD

**Issue by skinnider** [skinniderlab/CLM#247]
Here is a new error in `calculate_outcomes` (log is at /Genomics/argo/users/ms0270/git/CLM.snakemake/slurm_logs/rule_calculate_outcomes/6059528.log): 

```
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Provided resources: mem_mb=256000, mem_mib=244141, disk_mb=1000, disk_mib=954
Conda environments: ignored
Select jobs to execute...
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Provided resources: mem_mb=256000, mem_mib=244141, disk_mb=1000, disk_mib=954
Conda environments: ignored
Select jobs to execute...

[Tue Sep  3 23:13:12 2024]
rule calculate_outcomes:
    input: /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/structures_SMILES_2_unique_masses.csv.gz, /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/known_structures_SMILES_2_unique_masses.csv.gz, /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/invalid_structures_SMILES_2_unique_masses.csv.gz, /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/inputs/train0_structures_SMILES_2.smi
    output: /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/model_evaluation/30/structures_SMILES_2_calculate_outcomes.csv.gz
    jobid: 0
    reason: Forced execution
    wildcards: output_dir=/Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1, enum_factor=30, dataset=structures, repr=SMILES, fold=2
    resources: mem_mb=256000, mem_mib=244141, disk_mb=1000, disk_mib=954, tmpdir=/tmp, slurm_partition=main,hoppertest,skinniderlab, runtime=1000

reading NP model ...
model in
(INFO) (__main__.py) (03-Sep-24 23:13:17) CLM vsrc
(WARNING) (functions.py) (03-Sep-24 23:13:21) Not enough molecules for frequency bin '2-2'. Using 92868 molecules.
(WARNING) (functions.py) (03-Sep-24 23:13:21) Not enough molecules for frequency bin '3-10'. Using 88529 molecules.
(WARNING) (functions.py) (03-Sep-24 23:13:21) Not enough molecules for frequency bin '11-30'. Using 22206 molecules.
(WARNING) (functions.py) (03-Sep-24 23:13:21) Not enough molecules for frequency bin '31-100'. Using 9273 molecules.
(WARNING) (functions.py) (03-Sep-24 23:13:21) Not enough molecules for frequency bin '101-'. Using 33300 molecules.
(INFO) (calculate_outcomes.py) (03-Sep-24 23:13:23) Reading training smiles from /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/inputs/train0_structures_SMILES_2.smi
(INFO) (calculate_outcomes.py) (03-Sep-24 23:22:30) Reading sample smiles from                                                     smiles  ...   bin
0                Brc1cc(Br)c(Oc2c(Br)cc(Br)c(Br)c2Br)cc1Br  ...  3-10
1                  Brc1cc(Br)c(Oc2cc(Br)c(Br)cc2Br)c(Br)c1  ...  3-10
2                    Brc1cc(Br)c(Oc2cc(Br)c(Br)cc2Br)cc1Br  ...  3-10
3                      Brc1ccc(Oc2c(Br)cc(Br)cc2Br)c(Br)c1  ...   1-1
4                                             C#CC#CC#CCCO  ...  3-10
...                                                    ...  ...   ...
1246171                      O=C(O)CNC(=O)CCCCCCCCCc1ccco1  ...   all
1246172    C(C[N+]1(C=O)c2ccc(O)cc2)(CNc2nc(C(N)=O)nc2)CC1  ...   all
1246173                         C=C(C)C1CCC2(C)CCC(=O)C2C1  ...   all
1246174  C1(CO)OC(c2c3c(c(=O)cc(-c4ccc(O)cc4)o3)c(O)c2)...  ...   all
1246175                   CC1CCC2(O)C(C(C)(C)O)CC3CCCC12C3  ...   all

[1246176 rows x 5 columns]
 15%|█▌        | 159563/1030768 [30:14<2:49:02, 85.90it/s][23:52:45] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[23:52:45] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[23:52:45] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[23:52:45] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[23:52:45] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
 81%|████████▏ | 838053/1030768 [2:37:00<31:14, 102.83it/s][01:59:31] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[01:59:31] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[01:59:31] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[01:59:31] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[01:59:31] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
 99%|█████████▉| 1022461/1030768 [3:11:10<01:18, 106.08it/s][02:33:41] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[02:33:41] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[02:33:41] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[02:33:41] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
[02:33:41] WARNING: could not find number of expected rings. Switching to an approximate ring finding algorithm.
100%|██████████| 1030768/1030768 [3:22:13<00:00, 84.95it/s]
/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py:160: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  valid_df = pd.concat(valid_data.to_list())
/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py:160: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  valid_df = pd.concat(valid_data.to_list())
/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py:160: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  valid_df = pd.concat(valid_data.to_list())
/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py:160: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  valid_df = pd.concat(valid_data.to_list())
/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py:160: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  valid_df = pd.concat(valid_data.to_list())
/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py:160: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  valid_df = pd.concat(valid_data.to_list())
/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py:160: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  valid_df = pd.concat(valid_data.to_list())
/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py:160: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  valid_df = pd.concat(valid_data.to_list())
/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py:160: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  valid_df = pd.concat(valid_data.to_list())
/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py:160: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  valid_df = pd.concat(valid_data.to_list())
(INFO) (calculate_outcomes.py) (04-Sep-24 02:57:43) 1030768 valid SMILES out of 1246176
(INFO) (calculate_outcomes.py) (04-Sep-24 02:57:43) 1176066 novel SMILES out of 1246176
(INFO) (calculate_outcomes.py) (04-Sep-24 02:58:29) Calculating outcomes
(INFO) (calculate_outcomes.py) (04-Sep-24 02:58:31) Calculating outcomes for bin 1-1
(INFO) (calculate_outcomes.py) (04-Sep-24 03:52:56) Calculating outcomes for bin 101-
(INFO) (calculate_outcomes.py) (04-Sep-24 04:01:04) Calculating outcomes for bin 11-30
(INFO) (calculate_outcomes.py) (04-Sep-24 04:09:21) Calculating outcomes for bin 2-2
(INFO) (calculate_outcomes.py) (04-Sep-24 04:26:59) Calculating outcomes for bin 3-10
(INFO) (calculate_outcomes.py) (04-Sep-24 04:44:06) Calculating outcomes for bin 31-100
(INFO) (calculate_outcomes.py) (04-Sep-24 04:50:39) Calculating outcomes for bin all
Traceback (most recent call last):
  File "/Genomics/skinniderlab/PED-generation/env-clm/bin/clm", line 8, in <module>
    sys.exit(main())
  File "/Genomics/argo/users/ms0270/git/CLM/src/clm/__main__.py", line 81, in main
    args.func(args)
  File "/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py", line 346, in main
    calculate_outcomes(
  File "/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py", line 336, in calculate_outcomes
    out = calculate_outcomes_dataframe(sample_df, train_df)
  File "/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py", line 292, in calculate_outcomes_dataframe
    if _out := handle_bin(
  File "/Genomics/argo/users/ms0270/git/CLM/src/clm/commands/calculate_outcomes.py", line 192, in handle_bin
    np.concatenate(bin_df["elements"].to_numpy()), return_counts=True
ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 1 dimension(s) and the array at index 270620 has 0 dimension(s)
[Wed Sep  4 04:51:03 2024]
Error in rule calculate_outcomes:
    jobid: 0
    input: /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/structures_SMILES_2_unique_masses.csv.gz, /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/known_structures_SMILES_2_unique_masses.csv.gz, /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/invalid_structures_SMILES_2_unique_masses.csv.gz, /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/inputs/train0_structures_SMILES_2.smi
    output: /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/model_evaluation/30/structures_SMILES_2_calculate_outcomes.csv.gz
    conda-env: clm
    shell:
        clm calculate_outcomes --train_file /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/inputs/train0_structures_SMILES_2.smi --sampled_file /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/structures_SMILES_2_unique_masses.csv.gz --known_smiles_file /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/known_structures_SMILES_2_unique_masses.csv.gz --invalid_smiles_file /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/invalid_structures_SMILES_2_unique_masses.csv.gz --max_molecules 500000 --seed 12 --output_file /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/model_evaluation/30/structures_SMILES_2_calculate_outcomes.csv.gz
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
srun: error: argo-10: task 0: Exited with exit code 1
srun: launch/slurm: _step_signal: Terminating StepId=6059528.0
[Wed Sep  4 04:51:05 2024]
Error in rule calculate_outcomes:
    jobid: 0
    input: /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/structures_SMILES_2_unique_masses.csv.gz, /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/known_structures_SMILES_2_unique_masses.csv.gz, /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/invalid_structures_SMILES_2_unique_masses.csv.gz, /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/inputs/train0_structures_SMILES_2.smi
    output: /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/model_evaluation/30/structures_SMILES_2_calculate_outcomes.csv.gz
    conda-env: clm
    shell:
        clm calculate_outcomes --train_file /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/inputs/train0_structures_SMILES_2.smi --sampled_file /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/structures_SMILES_2_unique_masses.csv.gz --known_smiles_file /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/known_structures_SMILES_2_unique_masses.csv.gz --invalid_smiles_file /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/30/prior/samples/invalid_structures_SMILES_2_unique_masses.csv.gz --max_molecules 500000 --seed 12 --output_file /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/model_evaluation/30/structures_SMILES_2_calculate_outcomes.csv.gz
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
```

**Comment by vineetbansal** [comment]
TODO: Make sure that `canonical_smile` in `tabulate_molecues.py:62` can be passed on to `rdkit` to generate a valid molecule (one instance where it doesn't work is `"c1c2c(OC3C(O)C(O)C(O)C(C)O3)c3c4c(cc(C)oc(=O)c4c31)c1c(cc3c(C(C)(C=C)C)c4c(c(O)c1oc(=O)c2c(OC)c43)cccc4)C(=O)C=C(C)C4"`).

**Comment by skinnider** [comment]
Fixed in #248. For reference, command was:
```
snakemake --configfile /Genomics/skinniderlab/food-clm/clm/database=FooDB-representation=NA-metric=NA-optimizer=NA-rarefaction=1/config.yaml --slurm --jobs 50 --default-resources slurm_partition=main,hoppertest,skinniderlab --latency-wait=60 --rerun-incomplete
```
